File: src/main/java/net/ropes4k/impl/ConcatenationRope.java
/*
 * Copyright (C) 2024
 *  - Originally Copyright (C) 2007 Amin Ahmad.
 * Licenced under GPL
 */
package net.ropes4k.impl;

import java.io.IOException;
import java.io.Writer;
import java.util.Iterator;

import net.ropes4k.Rope;

/**
 * A rope that represents the concatenation of two other ropes.
 * @author Amin Ahmad
 */
public final class ConcatenationRope extends AbstractRope {

    private final Rope left;
    private final Rope right;
    private final byte depth;
    private final int length;

    /**
     * Create a new concatenation rope from two ropes.
     * @param left the first rope.
     * @param right the second rope.
     */
    public ConcatenationRope(Rope left, Rope right) {
        this.left   = left;
        this.right  = right;
        depth  = (byte) (Math.max(RopeUtilities.INSTANCE.depth(left), RopeUtilities.INSTANCE.depth(right)) + 1);
        length = left.length() + right.length();
    }

    @Override
    public char charAt(int index) {
        if (index >= length())
            throw new IndexOutOfBoundsException("Rope index out of range: " + index);

        return (index < left.length() ? left.charAt(index): right.charAt(index - left.length()));
    }

    @Override
    public byte depth() {
        return depth;
    }

    @Override
    public CharSequence getForSequentialAccess() {
        return getForSequentialAccess(this);
    }

    /*
     * Returns this object as a char sequence optimized for
     * regular expression searches.
     * <p>
     * <emph>This method is public only to facilitate unit
     * testing.</emph>
     */
    private CharSequence getForSequentialAccess(Rope rope) {
        return new CharSequence() {

            private final ConcatenationRopeIteratorImpl iterator = (ConcatenationRopeIteratorImpl) rope.iterator(0);

            @Override
            public char charAt(int index) {
                if (index > iterator.getPos()) {
                    iterator.skip(index- iterator.getPos()-1);
                    try {
                        return iterator.next();
                    } catch (IllegalArgumentException e) {
                        System.out.println("Rope length is: " + rope.length() + " charAt is " + index);
                        throw e;
                    }
                } else { /* if (index <= lastIndex) */
                    int toMoveBack = iterator.getPos() - index + 1;
                    if (iterator.canMoveBackwards(toMoveBack)) {
                        iterator.moveBackwards(toMoveBack);
                        return iterator.next();
                    } else {
                        return rope.charAt(index);
                    }
                }
            }

            @Override
            public int length() {
                return rope.length();
            }

            @Override
            public CharSequence subSequence(int start, int end) {
                return rope.subSequence(start, end);
            }

        };
    }

    /**
     * Return the left-hand rope.
     * @return the left-hand rope.
     */
    public Rope getLeft() {
        return left;
    }

    /**
     * Return the right-hand rope.
     * @return the right-hand rope.
     */
    public Rope getRight() {
        return right;
    }

    @Override
    public Iterator<Character> iterator(int start) {
        if (start < 0 || start > length())
            throw new IndexOutOfBoundsException("Rope index out of range: " + start);
        if (start >= left.length()) {
            return right.iterator(start - left.length());
        } else {
            return new ConcatenationRopeIteratorImpl(this, start);
        }
    }

    @Override
    public int length() {
        return length;
    }

    @Override
    public Rope rebalance() {
        return RopeUtilities.INSTANCE.rebalance(this);
    }

    @Override
    public Rope reverse() {
        return RopeUtilities.INSTANCE.concatenate(getRight().reverse(), getLeft().reverse());
    }

    @Override
    public Iterator<Character> reverseIterator(int start) {
        if (start < 0 || start > length())
            throw new IndexOutOfBoundsException("Rope index out of range: " + start);
        if (start >= right.length()) {
            return left.reverseIterator(start - right.length());
        } else {
            return new ConcatenationRopeReverseIteratorImpl(this, start);
        }
    }

    @Override
    public Rope subSequence(int start, int end) {
        if (start < 0 || end > length())
            throw new IllegalArgumentException("Illegal subsequence (" + start + "," + end + ")");
        if (start == 0 && end == length())
            return this;
        int l = left.length();
        if (end <= l)
            return left.subSequence(start, end);
        if (start >= l)
            return right.subSequence(start - l, end - l);
        return RopeUtilities.INSTANCE.concatenate(
            left.subSequence(start, l),
            right.subSequence(0, end - l));
    }

    @Override
    public void write(Writer out) throws IOException {
        left.write(out);
        right.write(out);
    }

    @Override
    public void write(Writer out, int offset, int length) throws IOException {
        if (offset + length <= left.length()) {
            left.write(out, offset, length);
        } else if (offset >= left.length()) {
            right.write(out, offset - left.length(), length);
        } else {
            int writeLeft = left.length() - offset;
            left.write(out, offset, writeLeft);
            right.write(out, 0, length - writeLeft);
        }
    }
}

File: src/main/kotlin/net/ropes4k/impl/ConcatenationRope.kt
/*
 * Copyright (C) 2024
 *  - Originally Copyright (C) 2007 Amin Ahmad.
 * Licenced under GPL
 */
package net.ropes4k.impl

import net.ropes4k.Rope
import net.ropes4k.impl.RopeUtilities.Companion.concatenate
import net.ropes4k.impl.RopeUtilities.Companion.depth
import net.ropes4k.impl.RopeUtilities.Companion.rebalance
import java.io.IOException
import java.io.Writer
import kotlin.math.max

/**
 * A rope that represents the concatenation of two other ropes.
 */
internal class ConcatenationRope(
    val left: Rope,
    val right: Rope
) : AbstractRope() {
    private val depth = max(depth(left), depth(right)) + 1


    override val length: Int = left.length + right.length

    override fun get(index: Int): Char {
        if (index >= length) throw IndexOutOfBoundsException("Rope index out of range: $index")

        return (if (index < left.length) left[index] else right[index - left.length])
    }

    override fun depth(): Int = depth

    /*
     * Returns this object as a char sequence optimized for
     * regular expression searches.
     * <p>
     */
    public override fun getForSequentialAccess(): CharSequence {
        return SequentialRopeAccessor(this)
    }

    class SequentialRopeAccessor(private val rope: Rope) : CharSequence {
        private val iterator = rope.iterator(0) as ConcatenationRopeIterator

        override fun get(index: Int): Char {
            if (index > iterator.pos) {
                iterator.skip(index - iterator.pos - 1)
                try {
                    return iterator.next()
                } catch (e: NoSuchElementException) {
                    throw IndexOutOfBoundsException("Index $index >= length $length")
                }
            } else { /* if (index <= lastIndex) */
                val toMoveBack = iterator.pos - index + 1
                if (iterator.canMoveBackwards(toMoveBack)) {
                    iterator.moveBackwards(toMoveBack)
                    return iterator.next()
                } else {
                    return rope[index]
                }
            }
        }

        override val length: Int get() = rope.length

        override fun subSequence(startIndex: Int, endIndex: Int): CharSequence {
            return rope.subSequence(startIndex, endIndex)
        }
    }

    override fun iterator(start: Int): Iterator<Char> {
        if (start < 0 || start > length) throw IndexOutOfBoundsException("Start $start < 0 or > length $length")
        return if (start >= left.length) {
            right.iterator(start - left.length)
        } else {
            ConcatenationRopeIterator(this, start)
        }
    }

    override fun rebalance(): Rope {
        return rebalance(this)
    }

    override fun reverse(): Rope {
        return concatenate(right.reverse(), left.reverse())
    }

    override fun reverseIterator(start: Int): Iterator<Char> {
        if (start < 0 || start > length) throw IndexOutOfBoundsException("Start $start < 0 or > length $length")
        return if (start >= right.length) {
            left.reverseIterator(start - right.length)
        } else {
            ConcatenationRopeReverseIterator(this, start)
        }
    }

    override fun subSequence(startIndex: Int, endIndex: Int): Rope {
        require(!(startIndex < 0 || endIndex > length)) { "Start/End ($startIndex/$endIndex) out of bounds (0/$length)" }
        if (startIndex == 0 && endIndex == length) return this
        val l = left.length
        if (endIndex <= l) return left.subSequence(startIndex, endIndex)
        if (startIndex >= l) return right.subSequence(startIndex - l, endIndex - l)
        return concatenate(
            left.subSequence(startIndex, l), right.subSequence(0, endIndex - l)
        )
    }

    @Throws(IOException::class)
    override fun write(out: Writer) {
        left.write(out)
        right.write(out)
    }

    @Throws(IOException::class)
    override fun write(out: Writer, offset: Int, length: Int) {
        if (offset + length <= left.length) {
            left.write(out, offset, length)
        } else if (offset >= left.length) {
            right.write(out, offset - left.length, length)
        } else {
            val writeLeft = left.length - offset
            left.write(out, offset, writeLeft)
            right.write(out, 0, length - writeLeft)
        }
    }
}
